### PR TITLE
System gtest

### PR DESCRIPTION
**Pull request by ellert**
# This Pull request:

## Changes or fixes:

Implement builtin_gtest option

By setting the option to OFF the system gtest and gmock are used.
This allows doing tests without network available, e.g. during a package build for Fedora/EPEL.

## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

### CODE REVIEW

**Comment by bellenot** [comment]
@ellert thanks for the PR, but it looks like it doesn't work...

**Comment by ellert** [comment]
> @ellert thanks for the PR, but it looks like it doesn't work...

The tests seem to be run with fail-on-missing.
Since most of the builtins have default OFF, I did the same for this one.
With fail-on-missing enabled the build then fails if the system googletest is not there.
So the test kind-of worked, since it tested this and acted accordingly.

Should I change the default value for the builtin_gtest to ON, or should the tests be changed to install the system googletest?
It probably makes most sense for you to change the default. I will try that.

**Comment by ellert** [comment]
That did not help. Even with the default set to ON, the CI tests are run with -Dbuiltin_gtest=Off, so they ignore the default.
It seems that the test definition must be changed too. Where are the tests defined? rootspi?
So for this to work in the CI testing it must be accompanied by a change in the definition of the CI tests to either call with -Dbuiltin_gtest=On (or at least not with an explicit -Dbuiltin_gtest=Off), or install the gtest-devel and gmock-devel packages in the test environment. Can someone give some pointers or help with the changes?

**Comment by bellenot** [comment]
@ellert I'll try to set up a VM with locally installed gtest-devel and gmock-devel packages, but I don't think we will change our CI 😉

**Comment by bellenot** [comment]
@phsft-bot build with flags -Dbuiltin_gtest=ON

**Comment by bellenot** [comment]
> That did not help. Even with the default set to ON, the CI tests are run with -Dbuiltin_gtest=Off, so they ignore the default. It seems that the test definition must be changed too. Where are the tests defined? rootspi? So for this to work in the CI testing it must be accompanied by a change in the definition of the CI tests to either call with -Dbuiltin_gtest=On (or at least not with an explicit -Dbuiltin_gtest=Off), or install the gtest-devel and gmock-devel packages in the test environment. Can someone give some pointers or help with the changes?

I think `builtin_gtest` should be added to the list of supported modules for different platforms in https://github.com/root-project/rootspi/blob/master/jenkins/root-build.cmake

**Comment by bellenot** [comment]
@ellert BTW, is it supposed to work on all platforms?

**Comment by ellert** [comment]
> @ellert BTW, is it supposed to work on all platforms?

For me it works for Fedora 34 - 37 and EPEL 7 - 9.
It should work for any system that has a libgtest and libgmock (.so or .a) in the system library path and headers in /usr/lib/gtest and /usr/lib/gmock.

I added a commit to the PR adding compatibility with the older gmock versions in EPEL for a recently added test.

**Comment by guitargeek** [comment]
Removing my request to review for now. Feel free to add me back in when the PR is in a state where the CI tests pass!

**Review by eguiraud — COMMENTED** [review]
LGTM for what regards the changes to `tree/dataframe/test`, I can't comment on the PR in general

**Comment by Axel-Naumann** [comment]
@phsft-bot build

File: math/mathcore/test/stress/testGenVector.cxx [hunk 1]
```diff
@@ -4,6 +4,16 @@
...
+#ifndef TYPED_TEST_SUITE_P
+#define TYPED_TEST_SUITE_P TYPED_TEST_CASE_P
+#endif
+#ifndef REGISTER_TYPED_TEST_SUITE_P
+#define REGISTER_TYPED_TEST_SUITE_P REGISTER_TYPED_TEST_CASE_P
+#endif
+#ifndef INSTANTIATE_TYPED_TEST_SUITE_P
+#define INSTANTIATE_TYPED_TEST_SUITE_P INSTANTIATE_TYPED_TEST_CASE_P
+#endif
+
```
Comment (Axel-Naumann) [inline math/mathcore/test/stress/testGenVector.cxx:16]: Can you add a comment `// Backward compatibility for gtest version < xyz`, please (with an appropriate "xyz") so we know in 10 years that we can safely remove this? (Same for the other similar occurrences.)

Comment (ellert) [reply math/mathcore/test/stress/testGenVector.cxx]: Done.

Comment (Axel-Naumann) [reply math/mathcore/test/stress/testGenVector.cxx]: Oh wow, thank you, super nice.

**Review by Axel-Naumann — APPROVED** [review]

**Pull request merged**